chainerx_cc/chainerx/cuda/cuda_device/batch_norm.cc

```diff
@@ -223,95 +282,104 @@ class CudaBatchNormForwardBackward : public chainerx::GenericBatchNormForwardBac
                 *x_desc,
                 internal::GetRawOffsetData(gout_cont),
                 *x_desc,
-                internal::GetRawOffsetData(gx),
+                internal::GetRawOffsetData(gx_cont),
                 *gamma_beta_mean_var_desc,
                 internal::GetRawOffsetData(gamma_casted_cont),
-                internal::GetRawOffsetData(ggamma),
-                internal::GetRawOffsetData(gbeta),
-                static_cast<double>(eps()),
+                internal::GetRawOffsetData(ggamma_casted_cont),
+                internal::GetRawOffsetData(gbeta_casted_cont),
+                static_cast<double>(eps),
                 internal::GetRawOffsetData(x_mean),
                 internal::GetRawOffsetData(x_inv_std));
 
-        if (ggamma.dtype() != gamma.dtype()) {
-            ggamma = ggamma.AsType(gamma.dtype());
+        const Array& actual_gx = gx.has_value() ? *gx : EmptyLike(x, device);
```

Comment:
It seems output arrays are allocated twice with `Empty`/`EmptyLike`.

Reply:
It should be fixed now can you take another look?

Reply:
Looks good.